smart_open/s3.py

```diff
@@ -0,0 +1,373 @@
+        return True
+
+    def seekable(self):
+        """If False, seek(), tell() and truncate() will raise IOError.
+
+        We offer only seek support, and no truncate support."""
+        return True
+
+    def seek(self, offset, whence=START):
+        """Seek to the specified position.
+
+        :param int offset: The offset in bytes.
+        :param int whence: Where the offset is from.
+
+        Returns the position after seeking."""
+        _LOGGER.debug('seeking to offset: %r whence: %r', offset, whence)
+        if whence not in WHENCE_CHOICES:
+            raise ValueError('invalid whence, expected one of %r' % WHENCE_CHOICES)
+
+        if whence == START:
+            new_position = offset
+        elif whence == CURRENT:
+            new_position = self._current_pos + offset
+        else:
+            new_position = self._content_length + offset
+        new_position = _clamp(new_position, 0, self._content_length)
+
+        _LOGGER.debug('new_position: %r', new_position)
+        self._current_pos = self._raw_reader.position = new_position
+        self._buffer = b""
+        self._eof = self._current_pos == self._content_length
+        return self._current_pos
+
+    def tell(self):
+        """Return the current position within the file."""
+        return self._current_pos
+
+    def truncate(self, size=None):
+        """Unsupported."""
+        raise io.UnsupportedOperation
+
+    #
+    # io.BufferedIOBase methods.
+    #
+    def detach(self):
+        """Unsupported."""
+        raise io.UnsupportedOperation
+
+    def read(self, size=-1):
+        """Read up to size bytes from the object and return them."""
+        if size <= 0:
+            if len(self._buffer):
+                from_buf = self._read_from_buffer(len(self._buffer))
+            else:
+                from_buf = b''
+            self._current_pos = self._content_length
+            return from_buf + self._raw_reader.read()
+
+        #
+        # Return unused data first
+        #
+        if len(self._buffer) >= size:
+            return self._read_from_buffer(size)
+
+        #
+        # If the stream is finished, return what we have.
+        #
+        if self._eof:
+            return self._read_from_buffer(len(self._buffer))
+
+        #
+        # Fill our buffer to the required size.
+        #
+        # _LOGGER.debug('filling %r byte-long buffer up to %r bytes', len(self._buffer), size)
+        while len(self._buffer) < size and not self._eof:
+            raw = self._raw_reader.read(size=io.DEFAULT_BUFFER_SIZE)
+            if len(raw):
+                self._buffer += raw
+            else:
+                _LOGGER.debug('reached EOF while filling buffer')
+                self._eof = True
+
+        return self._read_from_buffer(size)
+
+    def read1(self, size=-1):
+        """This is the same as read()."""
+        return self.read(size=size)
+
+    def readinto(self, b):
```

Comment:
Very strange, why we need this method with mutable `b`?

Reply:
Same as above. We try to behave like a [BufferedIOBase](https://docs.python.org/2/library/io.html#io.BufferedIOBase.readinto), which defines this method.

Some application downstream of smart_open may depend on this functionality.